### PR TITLE
Refs #28535 -- Add rebase option in makemigrations command

### PR DESCRIPTION
**Pull request by c-bata**
Hi! I added `--rebase` option for `makemigrations` command.
Usage is like this:

```
$ ls snippets/migrations/
0001_initial.py            0002_snippet_created_at.py 0002_snippet_updated_at.py __init__.py
$ python manage.py makemigrations --rebase -v 3
Rebasing snippets
  Branch 0002_snippet_updated_at
    - Add field updated_at to snippet
  Branch 0002_snippet_created_at
    - Add field created_at to snippet

Rebasing will only work if the operations printed above do not conflict
with each other (working on different fields or models)
Do you want to rebase these migration branches? [y/N] y

Created new rebase migrations:
  - /Users/a14737/PycharmProjects/django-book/tutorial/blog/snippets/migrations/0003_snippet_created_at.py
$ ls snippets/migrations/
0001_initial.py            0002_snippet_updated_at.py 0003_snippet_created_at.py __init__.py
```

Ticket URL is https://code.djangoproject.com/ticket/28535.
Thanks.

### CODE REVIEW

File: tests/migrations/test_commands.py [hunk 1]
```diff
@@ -898,6 +944,28 @@ def test_makemigrations_interactive_accept(self):
...
+    def test_makemigrations_interactive_accept_when_rebasing(self):
+        """
+        makemigrations --rebase enters interactive mode and rebase properly.
+        """
+        # Monkeypatch interactive questioner to auto accept
+        with mock.patch('builtins.input', mock.Mock(return_value='y')):
+            out = io.StringIO()
+            with self.temporary_migration_module(module="migrations.test_migrations_conflict") as migration_dir:
+                call_command("makemigrations", "migrations", rebase=True, interactive=True, stdout=out)
+                # makemigrations --rebase does not guarantee any particular order of the rebased migrations.
```
Comment (c-bata) [inline tests/migrations/test_commands.py]: I wonder if I should provide something argument to decide which migration files are rebased. But I don't have nice idea.

File: django/core/management/commands/makemigrations.py [hunk 2]
```diff
@@ -299,3 +315,122 @@ def all_items_equal(seq):
...
+                    MigrationAutodetector.parse_number(migration.name)
+                    for migration in rebase_migrations
+                ]
+                try:
+                    biggest_number = max(x for x in numbers if x is not None)
+                except ValueError:
+                    biggest_number = 1
+
+                # Get a path to the directory of migration files.
+                module_name, explicit = MigrationLoader.migrations_module(app_label)
```
Comment (c-bata) [inline django/core/management/commands/makemigrations.py]: Probably these code to know directory of migration files are not cool. But I don't know a better way.

**Comment by c-bata** [comment]
If this will be merged, it seems that the documentation should be updated.
So please tell me if you are positive to merge it. I'll try to update the documentation.
Thanks.

File: django/core/management/commands/makemigrations.py [hunk 3]
```diff
@@ -299,3 +315,122 @@ def all_items_equal(seq):
...
+                if module_name is None:
+                    continue
+                mod = import_module(module_name)
```
Comment (c-bata) [inline django/core/management/commands/makemigrations.py]: These lines are implemented by watching `django/db/migrations/loader.py` .
https://github.com/django/django/blob/master/django/db/migrations/loader.py#L64-L117

Perhaps it is possible to reduce these similar codes by refactoring.

**Comment by c-bata** [comment]
Could someone review this?

**Comment by ngnpope** [comment]
@c-bata You need to set the _has patch_ flag on the trac ticket to put it in the review queue. (See the [next steps](https://code.djangoproject.com/ticket/28535#ticket-next-steps) hint on the ticket.)

**Comment by c-bata** [comment]
@ngnpope Thanks. Sorry I didn't know about it.
I set `has patch` flag to true.

**Comment by OrDuan** [comment]
Any progress on this? What's the blocker?

**Comment by c-bata** [comment]
Thanks OrDuan. I don't know what's the blocker.
I hope someone review this.

**Comment by c-bata** [comment]
Hi @carltongibson.
Sorry I didn't mean to blame core developers and reviewers. I know that's hard to review this because django receives many pull requests. Thanks for always maintaining this great software.

> This is the main comment at this point. Without the documentation changes it's impossible to give a final review. Please add these changes.

OK. I'll update a documentation.

**Comment by c-bata** [comment]
Hi! It's been a while since I send this PR. Today I rebased this branch on master and re-check the diff. So this PR is ready for review.

**Comment by felixxm** [comment]
@c-bata Thanks for your huge effort :bowing_man:, unfortunately I don't think it's feasible to add `--rebase` with more complicated scenarios. I found many issues when testing only base use cases. This would be complicated and error-prone.

Thanks again :1st_place_medal: .

**Comment by c-bata** [comment]
I see. Thank you for your review!

**Pull request closed without merging**